src/librustc/infer/higher_ranked/mod.rs

```diff
@@ -520,6 +596,58 @@ pub fn leak_check<'a,'tcx>(infcx: &InferCtxt<'a,'tcx>,
             return Err((skol_br, tainted_region));
         }
     }
+
+    error_reporting::issue_32330_warnings(infcx, span, &warnings);
+
+    for (_, &skol) in skol_map {
+        // The outputs from a skolemized variable must all be
+        // equatable with `'static`.
+        let outgoing_taints = infcx.tainted_regions(snapshot,
+                                                    skol,
+                                                    TaintDirections::outgoing());
+        for &tainted_region in &outgoing_taints {
+            match tainted_region {
+                ty::ReVar(vid) if new_vars.contains(&vid) => {
+                    // Some region variable that doesn't escape
+                    // this snapshot:
+                    //
+                    //    [skol] -> [tainted_region]
+                    //
+                    // We can ignore this. We know that if it has
+                    // outgoing edges to any other regions, then those
+                    // regions must be 'static. If it has incoming
+                    // edges from other regions, we could therefore
+                    // infer it to 'static, and those edges would be
+                    // satisfied.
+                    //
+                    // We used to force such `tainted_region` to be
+                    // `'static`, but that leads to problems when
+                    // combined with `plug_leaks`. If you have a case
+                    // where `[skol] -> [tainted_region] -> [skol]`,
```

Comment:
So in that case we *really* can't force `[tainted_region]` to be `'static`, because it must be equal to `[skol]` which might not be `'static`.

My model is that we assign the tainted region the LUB of all regions below it (this can be `ReEmpty` if contravariance is involved - this is not WF but we Already Know That). Then we get a conflict if the LUB is above some region above it.

By the universal property of LUB, this is equivalent to all regions above being above of all regions below, so by induction only skolemized and external regions matter, and we already check these.

This has the Interesting Property of having `fn(&'static u8)` and `for<'a> fn(&'a u8)` be equivalent. This causes a coherence failure in:
```Rust
trait Marker {}

impl Marker for for<'a> fn(&'a u8) {}
impl Marker for fn(&'static u8) {}

fn main() {}
```

Hopefully this doesn't cause any headaches in trans due to eqty relationships changing.